Password Protection & Restrictions

adamaveray opened the pull request:
Hi there! Since this is an unofficial fork I don't know if you're taking feature additions, but I've imported and tidied up [this password protection extension](http://www.fpdf.org/en/script/script37.php) for the original FPDF library. (I couldn't make it a subclass since it needs to override a bunch of now-private methods.) It enables setting a password on the document, and prevent printing, editing, etc.

~~~php
$pdf = new tFPDF();
$pdf->SetProtection([$pdf::PERMISSION_PRINT, $pdf::PERMISSION_COPY], 'example-password');
~~~

adamaveray committed:
Supported protecting generated documents

kazsaj commented:
Hello Adam,

Sounds like a very useful addition, but I think it would be best to turn the private methods you need to overwrite to protected and implement that in a separate (encryption specific) subclass (just like you suggested).

Any chances you could make these changes and I will be happy to get this merged.

adamaveray committed:
Moved protected PDFs to subclass

adamaveray commented:
No worries, thanks for the feedback! I've moved it out to a subclass and updated the base class where necessary. I've tried to follow the existing code conventions but if I've missed anything just let me know and I'll update it.

kazsaj commented on src/tFPDF/ProtectedPDF.php:
```diff
@@ -0,0 +1,271 @@
+        $protection = self::PROTECTION_BASE;
+        foreach((array)$permissions as $permission){
+            $protection += $permission;
+        }
```
I think it would be better to do a binary check before adding each permission value. Just to prevent issues where the same flag is enabled twice, e.g.
```php
$pdf_instance->setProtection([$pdf_instance::PERMISSION_PRINT, $pdf_instance::PERMISSION_PRINT]);
```
Will actually add print twice, resulting in PERMISSION_MODIFY being enabled instead (but not print!).

I know this sounds like the developers is asking for trouble himself when passing duplicate values, but it's probably better to be safe than sorry, especially when dealing with such delicate things as permissions.

What I would suggest is:
```php
$protection = self::PROTECTION_BASE;
foreach ((array)$permissions as $permission){
    if (($protection & $permission) == 0) {
        $protection += $permission;
    }
}
```

kazsaj commented on src/tFPDF/ProtectedPDF.php:
```diff
@@ -0,0 +1,271 @@
+    const PERMISSION_COPY = 16;
+    const PERMISSION_ANNOTATE_FORMS = 32;
+    const PROTECTION_BASE = 192;
+    const PASSWORD_LENGTH = 32;
```
The constant probably should be called PASSWORD_MAX_LENGTH instead.

kazsaj commented on src/tFPDF/ProtectedPDF.php:
```diff
@@ -0,0 +1,271 @@
+    private function ObjectKey($object_index)
+    {
+        return substr($this->MD5to16($this->str_encryption_key.pack('VXxx', $object_index)), 0, 10);
+    }
```
For the sake of consistency I think it would be better to call the parameter $objectIndex

kazsaj commented on src/tFPDF/ProtectedPDF.php:
```diff
@@ -0,0 +1,271 @@
+    private static function GenerateUvalue($str_encryption_key)
+    {
+        return self::RC4($str_encryption_key, self::ENCRYPTION_PADDING);
+    }
```
Similar point here about naming convention - $encryptionKey would be more suitable.

kazsaj commented on src/tFPDF/ProtectedPDF.php:
```diff
@@ -0,0 +1,271 @@
+    {
+        $tmp = self::MD5to16($userPassword.$str_o_value.chr($protection)."\xFF\xFF\xFF");
+        return substr($tmp, 0, 5);
+    }
```
Last one about variable names - `$str_o_value` => `$oValue`

kazsaj reviewed (commented):
Left a few comments, hopefully I am not too annoying with my feedback.

adamaveray committed:
Handled duplicate permissions

adamaveray committed:
Fixed variable naming inconsistencies

adamaveray commented:
Thanks for going through that. I initially had all the variables camel-case, but noticed the base PDF class has everything snake-cased (with the type prefixes too ¯\\\_(ツ)\_/¯), so I started switching over to match that and I guess I got sidetracked! I've switched the remainder to snake-case to match the rest of the project, and addressed the other points you brought up. Let me know if you see any other issues!

kazsaj commented:
Thanks a lot! This is now merged and a new release was created after the merge:
https://github.com/DocnetUK/tfpdf/releases/tag/v2.1.0
